src/v/iceberg/schema_protobuf.cc

```diff
@@ -0,0 +1,185 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "iceberg/schema_protobuf.h"
+
+#include "iceberg/conversion_outcome.h"
+#include "iceberg/datatypes.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <fmt/core.h>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+#include <google/protobuf/descriptor.h>
+#include <google/protobuf/descriptor.pb.h>
+
+#include <deque>
+
+namespace iceberg {
+
+namespace {
+
+namespace pb = google::protobuf;
+
+/**
+ * Schema conversion from protocol buffers to internal Redpanda iceberg schema
+ */
+using field_outcome = conversion_outcome<nested_field_ptr>;
+using struct_outcome = conversion_outcome<struct_type>;
+using type_stack = std::deque<const pb::Descriptor*>;
+static constexpr int max_recursion_depth = 64;
+
+field_outcome
+from_protobuf(const pb::FieldDescriptor& fd, bool is_repeated, type_stack&);
+
+field_outcome success(const pb::FieldDescriptor& fd, field_type ft) {
+    return nested_field::create(
+      fd.number(), fd.name(), field_required::no, std::move(ft));
+}
+
+bool is_recursive_type(const pb::Descriptor& msg, type_stack& stack) {
+    return std::any_of(
+      stack.begin(), stack.end(), [&](const pb::Descriptor* d) {
+          return d == &msg;
```

Comment:
How comfortable are we that this pointers are always the same? Should we use `full_name` or similar?

Reply:
i assumed that descriptors are managed by the same `DescriptorPool` so they should point to the same type instance, i think the `full_name` is safer, will change that.